docs/errors.md

```diff
@@ -0,0 +1,231 @@
+- [`MongoServerError`](#MongoServerError)
+- [`MongoSystemError`](#MongoSystemError)
+
+## `MongoDriverError`
+
+This class represents errors which originate in the driver itself or in the user's use of the driver.
+This class should **never** be directly instantiated.
+Its children are the main classes of errors that most users will interact with: [**`MongoLogicError`**](#MongoLogicError) and [**`MongoRuntimeError`**](#MongoRuntimeError).
+
+### `MongoLogicError`
+
+This class represents errors which originate from misuse of the driver API and will generally be thrown before making contact with the server.
+This class should **never** be directly instantiated.
+![(MongoLogicError hierarchy tree)](charts/imgs/MongoLogicError.svg)
+Children of `MongoLogicError` include:
+
+- #### `MongoInvalidArgumentError`
+
+  - Thrown when the user supplies malformed, unexpected arguments or failed to provide a required argument or field.
+
+- #### `MongoCompatibilityError`
+
+  - Thrown when a feature that is not enabled or allowed for the current server configuration is used.
+
+- #### `MongoMissingCredentialsError`
+
+  - Thrown when a user fails to provide authentication credentials before attempting to connect to the mongo server.
+
+- #### `MongoMissingDependencyError`
+  - Thrown when a required module or dependency is not present.
+
+### `MongoRuntimeError`
+
+This class represents errors which occur when the driver encounters unexpected input or reaches an unexpected/invalid internal state.
+This class should **never** be directly instantiated.
+
+_MongoRuntimeError children (pt 1)_
+![(MongoRuntimeError hierarchy tree part 1)](charts/imgs/MongoRuntimeError_0.svg)
+
+_MongoRuntimeError children (pt 2)_
+![(MongoRuntimeError hierarchy tree part 2)](charts/imgs/MongoRuntimeError_1.svg)
+
+_MongoRuntimeError children (pt 3)_
+![(MongoRuntimeError hierarchy tree part 3)](charts/imgs/MongoRuntimeError_2.svg)
+
+Children of `MongoRuntimeError` include:
+
+- #### `MongoTransactionError`
+
+  - Thrown when the user makes a mistake in the usage of transactions (e.g.: attempting to commit a transaction with a readPreference other than primary).
+
+- #### `MongoNotConnectedError`
+
+  - Thrown when the user attempts to operate on the data from a client that has not been connected to a MongoDB server instance.
+
+- #### `MongoKerberosError`
+
+  - Thrown when the user attempts to authenticate via Kerberos, but fails to connect to the Kerberos client.
+
+- #### `MongoEncryptionError`
+
+  - Thrown when an autoencrypter is requested but unavailable.
+
+- #### `MongoCompressionError`
+
+  - Thrown when the driver fails to compress data before sending it to the server.
+
+- #### `MongoDecompressionError`
+
+  - Thrown when the driver fails to decompress data received from the server
+
+- #### `MongoExpiredSessionError`
+
+  - Thrown when the user attempts to operate on a session that has expired or has been closed.
+
+- #### `MongoParseError`
```

Comment:
I think this might also be making more explicit, e.g.`MongoURIParseError`, thoughts?

Reply:
I've renamed `MongoParseError` to `MongoURIError` to be more specific in what it's targeting while still being able to expand this to cases where the user supplies a properly formatted URI, but it contains illegal options.